zerver/lib/actions.py

```diff
@@ -818,14 +818,54 @@ def do_set_realm_property(
     """Takes in a realm object, the name of an attribute to update, the
     value to update and and the user who initiated the update.
     """
-    property_type = Realm.property_types[name]
-    assert isinstance(
-        value, property_type
-    ), f"Cannot update {name}: {value} is not an instance of {property_type}"
 
-    old_value = getattr(realm, name)
-    setattr(realm, name, value)
-    realm.save(update_fields=[name])
+    # We want to make sure all the db changes happen atomically,
+    # so we have to split up the logic to first do its database-modification
+    # part, and then, after exiting the atomic block, take care of the
+    # remaining side-effects such as flushing cashes and sending events.
+    # They shouldn't be done in the atomic block to avoid holding prolonged db locks
+    # if they're stuck or slow for any reason.
```

Comment:
Did you consider doing a `transaction.on_commit` for these side effects?  I think that would feel more readable, since then we could keep the event updates immediately adjacent to the actual changes.  

Possibly doing that would let us to just do `@transaction.atomic()` on the function itself rather than indenting everything, which might be a better pattern for these more generally.

Reply:
@timabbott Yeah, that can be done. I feel that splitting up the logic instead of `on_commit` is a bit simpler in the end because `on_commit` increases complexity in tests - you have to use `captureOnCommitCallbacks` in them to have the side-effects happen, and tests that don't use it but have these functions executed somewhere down the tested codepath live in a weird world where those side-effects didn't happen.

On the other hand, using `on_commit` across the board will let us to use `atomic()` on functions that should have it but currently can't because they have the structure:
```
def tricky_function_that_should_be_atomic():
  # a bunch of db operations
  another_function_that_should_be_executed_in_the_same_transaction_but_calls_send_event()
``` 
e.g. `do_deactivate_realm - calls send_event deep down the codepath in downgrade billing logic`.

So I think there's a balance here - `on_commit` makes using `transaction.atomic` easier, but if it's overdone, tests can diverge from reality a bit too much :thinking: 

@abhijeetbodas2001 Do you have any additional thoughts on this? Since you've worked on adding `on_commit` in some places already.

Reply:
I do remember testing that we've put `on_commit` being a real pain. So, unless we can arrange for any linting rules etc to prevent someone from forgetting using `on_commit` for events, `with transaction.atomic()` with the indent is definitely better from a testing POV. But the case @mateuszmandera describes does seem really problematic, so I'm confused now too :sweat_smile:

I remember there being some discussion on https://chat.zulip.org/ about moving from RabbitMQ to PostgreSQL queues? If that transition is in the plans, I guess this would become a non-issue?

Reply:
> I remember there being some discussion on https://chat.zulip.org/ about moving from RabbitMQ to PostgreSQL queues? If that transition is in the plans, I guess this would become a non-issue?

Hmm, interesting I haven't thought about that, but quite possibly it would? Not sure how likely that transition is to happen though

Reply:
> I remember there being some discussion on https://chat.zulip.org/ about moving from RabbitMQ to PostgreSQL queues? If that transition is in the plans, I guess this would become a non-issue?

Hmm, interesting I didn't think of that before, but quite possibly? Not sure how likely that transition is to happen though

Reply:
I thought that we resolved the testing issues with using `on_commit` via the shared logic in `tornado_redirected_to_list`, which is used for all of our tests that verify `send_event` logic?  I think there's no longer anything special required to test code that is using `transaction.on_commit`.

Reply:
Yeah, it seems to solve it - the downside is the tests need to know they have to call it - but as long as this only needs to be thought about when verifying events, perhaps its mine. It would perhaps get trickier if other kinds of side-effects end up in `on_commit`, such as email sending, external http requests etc.? Which may be entirely avoidable.

Reply:
Yeah, I think for now we should only be committed to doing the `send_event` calls in `transaction.atomic`; in practice we don't have a lot of `actions` events with HTTP calls/emails anyway.

Reply:
> I thought that we resolved the testing issues with using on_commit via the shared logic in tornado_redirected_to_list, which is used for all of our tests that verify send_event logic?

That's right. I meant to say that, as far as I know, there is no easy way currently to verify (in tests) *that we have used `on_commit`* for events inside transactions. Testing the events themselves is not what I meant.

This is one test which I had written to check that `on_commit` *is* used. Unless we have a linting rule like "Don't allow bare `send_event` call inside the scope of a transaction", we will need tests like this for every use of `on_commit` right?
```py
    # From zerver/tests/test_reactions.py
    def test_events_sent_after_transaction_commits(self) -> None:
        """
        Tests that `send_event` is hooked to `transaction.on_commit`. This is important, because
        we don't want to end up holding locks on message rows for too long if the event queue runs
        into a problem.
        """
        hamlet = self.example_user("hamlet")
        self.send_stream_message(hamlet, "Denmark")
        message = self.get_last_message()
        reaction = Reaction(
            user_profile=hamlet,
            message=message,
            emoji_name="whatever",
            emoji_code="whatever",
            reaction_type="whatever",
        )

        with self.tornado_redirected_to_list([], expected_num_events=1):
            with mock.patch("zerver.lib.actions.send_event") as m:
                m.side_effect = AssertionError(
                    "Events should be sent only after the transaction commits."
                )
                notify_reaction_update(hamlet, message, reaction, "stuff")
```

Reply:
Ultimately, once we finish this sweep, we can at least confirm every `send_event` call uses `transaction.on_commit` with an audit, and then potentially add a `semgrep` lint rule.